lib/rust/parser/src/syntax/statement.rs

```diff
@@ -334,19 +341,164 @@ fn parse_statement<'s>(
             }
         }
         Some(_) => unreachable!(),
-        None => {
-            let statement = precedence.resolve_offset(start, items);
-            debug_assert!(items.len() <= start);
-            Line {
-                newline,
-                content: apply_private_keywords(
-                    statement,
-                    items.drain(..),
-                    statement_context.visibility_context,
-                )
-                .map(StatementOrPrefix::Statement),
-            }
+        None => parse_expression_statement(
+            prefixes,
+            start,
+            item::Line { newline, items: mem::take(items) },
+            precedence,
+            statement_context.visibility_context,
+        )
+        .map_content(StatementOrPrefix::Statement),
+    }
+}
+
+fn take_trailing_newlines<'s>(
+    prefixes: &mut Vec<Line<'s, StatementPrefix<'s>>>,
+    trailing_newlines: &mut usize,
+    first_newline: &mut token::Newline<'s>,
+) -> Vec<token::Newline<'s>> {
+    let mut newlines = Vec::with_capacity(1 + *trailing_newlines);
+    newlines.extend((0..*trailing_newlines).map(|_| {
+        let Some(Line { newline, content: None }) = prefixes.pop() else { unreachable!() };
+        mem::replace(first_newline, newline)
+    }));
```

Comment:
Worth mentioning in the method, that only additional newlines are returned: the first one is returned separately.

Or maybe just make the code more readable? I have impression this `Vec` creation + `extend` could be something like `let newlines = prefixes.drain(prefixes.len() - trailing_newlines)`
Then the first element from iterator goes to `first_newline` and the rest may be collected.

Reply:
Added a `StatementPrefixes` type providing a higher-level interface.